File: app/services/invitation_rendering/font_registry.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any


REGISTRY_PATH = Path(__file__).resolve().parents[3] / "shared" / "font-registry.json"


@lru_cache(maxsize=1)
def _registry() -> dict[str, Any]:
    return json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def _family_by_alias() -> dict[str, str]:
    aliases: dict[str, str] = {}
    for font in _registry()["families"]:
        family = str(font["family"])
        for name in (family, *font.get("aliases", [])):
            aliases[str(name).strip().casefold()] = family
    return aliases


def _first_css_family(value: object) -> str:
    family = str(value or "").split(",", 1)[0].strip().strip("\"'").strip()
    return family


def normalize_font_family(value: object) -> str:
    requested = _first_css_family(value).casefold()
    return _family_by_alias().get(requested, str(_registry()["defaultFamily"]))


def bundled_font_families() -> tuple[str, ...]:
    return tuple(str(font["family"]) for font in _registry()["families"])

File: frontend/lib/flyer/fontRegistry.ts
import registry from "@/lib/flyer/font-registry.json";

export type BundledFontCategory = "sans" | "serif" | "handwriting" | "display";

export interface BundledFontFamily {
  key: string;
  family: string;
  category: BundledFontCategory;
  aliases: string[];
  faces: {
    normal: Record<"400" | "700", string>;
    italic: Record<"400" | "700", string>;
  };
}

export const DEFAULT_FLYER_FONT_FAMILY = registry.defaultFamily;
export const BUNDLED_FONT_FAMILIES = registry.families as BundledFontFamily[];
export const BUNDLED_FONT_FAMILY_NAMES = BUNDLED_FONT_FAMILIES.map(
  ({ family }) => family,
);

const FAMILY_BY_ALIAS = new Map<string, BundledFontFamily>();
for (const font of BUNDLED_FONT_FAMILIES) {
  for (const name of [font.family, ...font.aliases]) {
    FAMILY_BY_ALIAS.set(name.trim().toLowerCase(), font);
  }
}

function firstCssFamily(value: unknown): string {
  return String(value ?? "")
    .split(",", 1)[0]
    .trim()
    .replace(/^["']|["']$/g, "")
    .trim();
}

export function getBundledFont(value: unknown): BundledFontFamily {
  const requested = firstCssFamily(value).toLowerCase();
  return FAMILY_BY_ALIAS.get(requested)
    ?? BUNDLED_FONT_FAMILIES.find(
      ({ family }) => family === DEFAULT_FLYER_FONT_FAMILY,
    )!;
}

export function normalizeFontFamily(value: unknown): string {
  return getBundledFont(value).family;
}

export function normalizeFlyerFontConfiguration<
  T extends { stub_guest_font_family?: unknown },
>(configuration: T): T & { stub_guest_font_family: string } {
  return {
    ...configuration,
    stub_guest_font_family: normalizeFontFamily(
      configuration.stub_guest_font_family,
    ),
  };
}

export function bundledFaceWeight(weight: unknown): "400" | "700" {
  const normalized = String(weight ?? "normal").toLowerCase();
  return normalized === "bold" || normalized === "semibold"
    || Number(normalized) >= 600
    ? "700"
    : "400";
}
